pkg/server/preprocessor/preprocessor.go

```diff
@@ -50,16 +51,29 @@ func isTokenStillValid(token router.AccessToken, userInfo skydb.UserInfo) bool {
 
 func (p InjectUserIfPresent) Preprocess(payload *router.Payload, response *router.Response) int {
 	if payload.UserInfoID == "" {
-		log.Debugln("injectUser: empty UserInfoID, skipping")
-		return http.StatusOK
+		if !payload.HasMasterKey() {
+			log.Debugln("injectUser: empty UserInfoID, skipping")
+			return http.StatusOK
+		}
+		payload.UserInfoID = "_god"
+		payload.Context = context.WithValue(payload.Context, router.UserIDContextKey, "_god")
 	}
 
 	conn := payload.DBConn
 	userinfo := skydb.UserInfo{}
 	if err := conn.GetUser(payload.UserInfoID, &userinfo); err != nil {
-		log.Errorf("Cannot find UserInfo.ID = %#v\n", payload.UserInfoID)
-		response.Err = skyerr.NewError(skyerr.UnexpectedUserInfoNotFound, err.Error())
-		return http.StatusInternalServerError
+		if payload.HasMasterKey() {
```

Comment:
It appears that you are assuming that `GetUser` returns an error iff the user is not found. In reality an error could also occurs for other kinds of error such as other database problems. I think it is better to check the error type to make sure the user does not exist before creating a new user.